pkg/retriever/cidretrieval.go

```diff
@@ -0,0 +1,502 @@
+			retrievalErrors = multierr.Append(retrievalErrors, ErrAllRetrievalsFailed)
+		}
+		return nil, multierr.Append(queryErrors, retrievalErrors)
+	}
+	// if we succeeded, drop all the errors that occurred along the way
+	return stats, nil
+}
+
+// runRetrieval is a singular CID:SP retrieval, expected to be run in a goroutine
+// and coordinate with other candidate retrievals to block after query phase and
+// only attempt one retrieval-proper at a time.
+func (retrieval *cidRetrieval) runRetrieval(ctx context.Context, state *retrievalState, candidate RetrievalCandidate) {
+	queryResponse, err := queryCandidate(ctx, retrieval.GetStorageProviderTimeout, retrieval.Client, state, candidate)
+	if err != nil {
+		if err != errQueryCancelled {
+			retrieval.Instrumentation.OnErrorQueryingRetrievalCandidate(candidate, err)
+			state.sendResult(runResult{QueryError: err})
+		}
+		return
+	}
+
+	retrieval.Instrumentation.OnRetrievalQueryForCandidate(candidate, queryResponse)
+
+	if queryResponse.Status != retrievalmarket.QueryResponseAvailable ||
+		!retrieval.IsAcceptableQueryResponse(queryResponse) {
+		// bail, with no result or error
+		state.sendResult(runResult{})
+		return
+	}
+
+	retrieval.Instrumentation.OnFilteredRetrievalQueryForCandidate(candidate, queryResponse)
+
+	// priorityqueue wait; gated here so that only one retrieval can happen at once
+	done := state.WaitQueue.Wait(queryResponse)
+	defer done()
+
+	if !state.canSendResult() {
+		// a retrieval already successfully finished, don't proceed
+		return
+	}
+
+	retrieval.Instrumentation.OnRetrievingFromCandidate(candidate)
+
+	log.Infof(
+		"Attempting retrieval from miner %s for %s",
+		candidate.MinerPeer.ID,
+		formatCidAndRoot(state.Cid, candidate.RootCid, false),
+	)
+
+	stats, err := retrieveFromCandidate(ctx, retrieval.GetStorageProviderTimeout, retrieval.Client, candidate, queryResponse)
+
+	if err != nil {
+		retrieval.Instrumentation.OnErrorRetrievingFromCandidate(candidate, err)
+		state.sendResult(runResult{RetrievalError: err})
+	} else {
+		state.sendResult(runResult{RetrievalResult: stats})
+	}
+}
+
+// canSendResult will indicate whether a result is likely to be accepted (true)
+// or whether the retrieval is already finished (likely by a success)
+func (state *retrievalState) canSendResult() bool {
+	select {
+	case <-state.FinishChan:
+		return false
+	default:
+	}
+	return true
+}
+
+// sendResult will only send a result to the parent goroutine if a retrieval has
+// finished (likely by a success), otherwise it will send the result
+func (state *retrievalState) sendResult(result runResult) bool {
+	select {
+	case <-state.FinishChan:
+		return false
+	case state.ResultChan <- result:
+		if result.RetrievalResult != nil {
+			// signals to goroutines to bail, this has to be done here, rather than on
+			// the receiving parent end, because immediately after this call we instruct
+			// the prioritywaitqueue that we're done and another may start
+			close(state.FinishChan)
```

Comment:
Mmmm, at first glance this goes against what I've read about closing channels from N senders. Let's talk about this.

Reply:
Oh I see, you're relying on the priority queue named WaitQueue to ensure you only do one retrieval at a time.

Reply:
yeah, I move this here because my tests showed a case where there's a small gap between the waitqueue being told we're done with one, and the result being received, allowing the next one to pick it up; so the `close()` can only happen while the rest are blocked